### PR TITLE
Unit tests for viewRecords

### PR DESCRIPTION
**Pull request by hannespetur**
Pull request within a pull request. A pullception (copyright under consideration).

Several unit tests for `viewRecords` added. This change also adds a few additional tests for the `jumpToRegion` function, which makes sure the correct read is "jumped" to when the function is used.

Best,
Hannes

### CODE REVIEW

**Comment by hannespetur** [comment]
@smehringer the tests for viewRecords are not passing right now, but I am not sure why :astonished:  I will try run `gdb` to see what is going on..

I will rebase my branch if you make any changes so don't worry about creating conflicts.

Best wishes,
Hannes

**Comment by hannespetur** [comment]
Nevermind, the failed test was due to a bug in one of the tests. Do you think these unit tests will cover enough @smehringer ?

File: include/seqan/bam_io/bam_index_bai.h [hunk 1]
```diff
@@ -61,6 +61,9 @@
 #ifndef INCLUDE_SEQAN_BAM_IO_BAM_INDEX_BAI_H_
 #define INCLUDE_SEQAN_BAM_IO_BAM_INDEX_BAI_H_
 
+#include <cstdint>
```
Comment (smehringer) [inline include/seqan/bam_io/bam_index_bai.h]: why this?

File: tests/bam_io/test_bam_index.h [hunk 2]
```diff
@@ -84,15 +84,30 @@ SEQAN_DEFINE_TEST(test_bam_io_bam_index_open)
...
+    SEQAN_ASSERT(jumpToRegion(bamFile, hasAlignments, 0, 3, 100, baiIndex));
+    SEQAN_ASSERT_MSG(hasAlignments, "There should be alignments in region [3,100].");
+    readRecord(record, bamFile);
+    SEQAN_ASSERT_MSG(record.beginPos == 2, "Jumping to region [3,10] should go to the third read.");
+    SEQAN_ASSERT(atEnd(bamFile));
+
+    SEQAN_ASSERT_NOT(jumpToRegion(bamFile, hasAlignments, 1, 1, 10, baiIndex));
+    SEQAN_ASSERT_NOT_MSG(hasAlignments, "No alignments should be on contig 1.");
+
+    SEQAN_ASSERT(jumpToRegion(bamFile, hasAlignments, 0, 4, 100, baiIndex));
```
Comment (smehringer) [inline tests/bam_io/test_bam_index.h]: The tests always have a read starting exactly at the beginning of the region. Maybe we add one more read at pos... 42? and jump to region `[4,100]` expecting success.

File: tests/bam_io/test_bam_index.h [hunk 3]
```diff
@@ -109,5 +124,79 @@ SEQAN_DEFINE_TEST(test_bam_io_bam_index_save)
...
+
+    BamIndex<Bai> baiFile;  // Open BAI index file
+    open(baiFile, toCString(baiFileName));
+
+    BamHeader header;
+    readHeader(header, bamFile);
+
+    std::vector<BamAlignmentRecord> records;
+
+    viewRecords(records, bamFile, baiFile, 0, 1, 0);
```
Comment (smehringer) [inline tests/bam_io/test_bam_index.h:160]: Nice test case :+1: . we should add this for jumpToRegion

File: tests/bam_io/test_bam_index.h [hunk 4]
```diff
@@ -109,5 +124,79 @@ SEQAN_DEFINE_TEST(test_bam_io_bam_index_save)
...
+    SEQAN_ASSERT_MSG(length(records) == 0, "No reads expected in region seq1:1-0");
+    viewRecords(records, bamFile, baiFile, 0, 2000, 2000);
+    SEQAN_ASSERT_MSG(length(records) == 0, "No reads expected in region seq1:2000-2000");
+    viewRecords(records, bamFile, baiFile, 0, 2000, 100000);
+    SEQAN_ASSERT_MSG(length(records) == 0, "No reads expected in region seq1:2000-100000");
```
Comment (smehringer) [inline tests/bam_io/test_bam_index.h]: what is the difference to the test case before?

File: tests/bam_io/test_bam_index.h (hunk 3 above)
Comment (smehringer) [reply tests/bam_io/test_bam_index.h:160]: we should also add one with the region `[0, 100]` where it is clearly not 1-based.

**Review by smehringer — COMMENTED** [review]
Thank you so much for the tests! 

I think this covers a lot. I have some minor suggestions for a few more (see comments) and I will think about it on monday again.
:smiley:

**Comment by smehringer** [comment]
PS: better make that copyright fast :running_man:  :dash:

File: include/seqan/bam_io/bam_index_bai.h (hunk 1 above)
Comment (hannespetur) [reply include/seqan/bam_io/bam_index_bai.h]: Oops, I forgot to remove those... I have been trying out different IDEs lately and one gave me all these "Undefined type 'uint32_t' blahblah" errors without these two includes. I will remove them.

File: tests/bam_io/test_bam_index.h (hunk 2 above)
Comment (hannespetur) [reply tests/bam_io/test_bam_index.h]: Good point. There are some other tests that depend on the `small.bam` file that I don't want to break though.. I think I will rewrite the `jumpToRegion` tests using `ex1.bam` instead so I can more easily cover other cases.

File: tests/bam_io/test_bam_index.h (hunk 3 above)
Comment (hannespetur) [reply tests/bam_io/test_bam_index.h:160]: Agreed. But the region `[0, 100]` should throw an assertion failure, is it possible to check for that?

File: tests/bam_io/test_bam_index.h (hunk 4 above)
Comment (hannespetur) [reply tests/bam_io/test_bam_index.h]: None, really :smile:

**Comment by hannespetur** [comment]
Thank you for a fast response and a good review @smehringer! I think I have addressed all your remarks except for one that I wasn't sure how to do in SeqAn (add test with region `[0, 100]` that should fail).

**Comment by smehringer** [comment]
There is the `SEQAN_TEST_EXCEPTION_WITH_MESSAGE(...)` macro :)
I will add it, since I noticed that I only included a debug assertion but did not throw anything so far :innocent:

File: include/seqan/bam_io/bam_index_bai.h [hunk 5]
```diff
@@ -61,6 +61,7 @@
 #ifndef INCLUDE_SEQAN_BAM_IO_BAM_INDEX_BAI_H_
 #define INCLUDE_SEQAN_BAM_IO_BAM_INDEX_BAI_H_
 
+
```
Comment (smehringer) [inline include/seqan/bam_io/bam_index_bai.h:64]: :nail_care:

**Comment by smehringer** [comment]
@hannespetur Do you think we should also throw an assertion if the specified region end position is larger than the actual contig size? I can check for this.. but i don't know if an exception should be thrown or if the region end should simply be put to the length of the contig. (I have a developer in mind doesn't check for region boundaries based on an user input. So the user just wants a tool to run through, but because of the careless developer the programm throws an assertion... :thinking: )

**Comment by hannespetur** [comment]
IMO it is nicer if the region end should be put as the length of the contig instead of throwing an exception. `samtools view` handles it like that that and I think it is fine. I think it is alright not to throw an exception when the user's intent is clear.

I also have noticed that `samtools view` parses regions starting at position 0 without an error, e.g. if passed chr21:0-10000000 it just assumes the user means chr21:1-10000000. Maybe that case should not throw either.

File: include/seqan/bam_io/bam_index_bai.h (hunk 5 above)
Comment (hannespetur) [reply include/seqan/bam_io/bam_index_bai.h:64]: The file is clearly unreadable without that newline 😃

**Comment by smehringer** [comment]
Thanks for checking samtools! I agree we should not throw in this case but I will add debug assertions for both cases. It's more user friendly to not let the program abort.

**Comment by hannespetur** [comment]
Great, many thanks!

**Pull request merged**